cpu/kinetis_common/hwtimer_arch.c

```diff
@@ -47,23 +94,80 @@ static hwtimer_stimer32b_t stimer;
  */
 void (*timeout_handler)(int);
 
-inline static uint32_t lptmr_get_cnr(void)
-{
 #if LPTIMER_CNR_NEEDS_LATCHING
+
+/* read the CNR register */
+inline static uint16_t lptmr_get_cnr(void)
+{
+    uint16_t cnr;
     /* Write some garbage to CNR to latch the current value */
-    LPTIMER_DEV->CNR = 42;
-    return (uint32_t)LPTIMER_DEV->CNR;
-#else
+    asm volatile (
+        /* write garbage to CNR to latch the value, it does not matter what the
+         * value of r0 is. Testing shows that the write must be 32 bit wide or
+         * we will get garbage when reading back CNR. */
+        "str %[CNR], [%[CNR_p], #0]\n"
+        "ldrh %[CNR], [%[CNR_p], #0]\n" /* Read CNR */
+        : /* Output variables */
+        [CNR] "=&r" (cnr)
+        : /* Input variables */
+        [CNR_p] "r" (&(LPTIMER_DEV->CNR))
+        : /* Clobbered registers */
+        );
+    return cnr;
+}
+
+/* Update the CMR register and return the old CNR before the reset */
+inline static uint16_t lptmr_update_cmr(uint16_t cmr)
+{
+    /* Optimized version for CPUs which support latching the LPTMR value */
+    uint16_t cnr;
+    asm volatile (
+        /* Prepare On/Off values in registers r6, r7 */
+        "mov r6, #0\n"
+        "mov r7, #1\n"
+        /* Any ticks that occur between the below write to CNR_p and the write
+         * to TEN below will be lost, try to keep it short to minimize the
+         * risk. */
+        /* write garbage to CNR to latch the value, it does not matter what the
+         * value of r7 is. Testing shows that the write must be 32 bit wide or
+         * we will get garbage when reading back CNR. */
+        "str r7, [%[CNR_p], #0]\n"
+        "ldrh %[CNR], [%[CNR_p], #0]\n" /* Read CNR */
+        "strh r6, [%[TEN], #0]\n"       /* disable timer */
+        "strh %[CMR], [%[CMR_p], #0]\n" /* update CMR */
+        "strh r7, [%[TEN], #0]\n"       /* enable timer */
+        : /* Output variables */
+        [CNR] "=&r" (cnr)
+        : /* Input variables */
+        [TEN] "r" (BITBAND_ADDR(&(LPTIMER_DEV->CSR), LPTMR_CSR_TEN_SHIFT)),
```

Comment:
BITBAND_ADDR macro is not defined for k64f and kw2x and there is no advantage.

cpu/k60/include/cpu_conf.h:#define BITBAND_ADDR(addr, bit) ((((uint32_t) (addr)) & 0xF0000000u) + 0x2000000 + ((((uint32_t) (addr)) & 0xFFFFF) << 5) + ((bit) << 2))